presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java

```diff
@@ -198,21 +206,52 @@ else if (!info.getState().isDone()) {
                 .map(nodeStatus -> nodeStatus.getNodeStatus().getMemoryInfo())
                 .collect(toImmutableList());
 
-        Map<MemoryPoolId, Long> counts = nodeQueryStates.values().stream()
-                .map(CoordinatorQueriesState::getActiveQueries)
-                .flatMap(Collection::stream)
-                .collect(groupingBy(query -> query.getBasicQueryInfo().getMemoryPool(), counting()));
-        // Add defaults when queries are not running
-        counts.putIfAbsent(GENERAL_POOL, 0L);
+        int queriesAssignedToGeneralPool = 0;
+        int queriesAssignedToReservedPool = 0;
+        Query largestGeneralPoolQuery = null;
+        for (CoordinatorQueriesState nodeQueryState : nodeQueryStates.values()) {
+            for (Query query : nodeQueryState.getActiveQueries()) {
+                MemoryPoolId memoryPool = query.getBasicQueryInfo().getMemoryPool();
+                if (GENERAL_POOL.equals(memoryPool)) {
+                    queriesAssignedToGeneralPool = Math.incrementExact(queriesAssignedToGeneralPool);
+                    if (!resourceOvercommit(query.getBasicQueryInfo().getSession().toSession(sessionPropertyManager))) {
+                        largestGeneralPoolQuery = getLargestMemoryQuery(largestGeneralPoolQuery, query);
+                    }
+                }
+                else if (RESERVED_POOL.equals(memoryPool)) {
+                    queriesAssignedToReservedPool = Math.incrementExact(queriesAssignedToReservedPool);
+                }
+                else {
+                    throw new IllegalArgumentException("Unrecognized memory pool: " + memoryPool);
+                }
+            }
+        }
+
+        ImmutableMap.Builder<MemoryPoolId, ClusterMemoryPoolInfo> memoryPoolInfos = ImmutableMap.builder();
+        ClusterMemoryPool pool = new ClusterMemoryPool(GENERAL_POOL);
+        pool.update(memoryInfos, queriesAssignedToGeneralPool);
+        ClusterMemoryPoolInfo clusterInfo = pool.getClusterInfo(Optional.ofNullable(largestGeneralPoolQuery).map(Query::getQueryId));
+        memoryPoolInfos.put(GENERAL_POOL, clusterInfo);
         if (isReservedPoolEnabled) {
-            counts.putIfAbsent(RESERVED_POOL, 0L);
+            pool = new ClusterMemoryPool(RESERVED_POOL);
+            pool.update(memoryInfos, queriesAssignedToReservedPool);
+            memoryPoolInfos.put(RESERVED_POOL, pool.getClusterInfo());
         }
+        return memoryPoolInfos.build();
+    }
 
-        return counts.entrySet().stream().collect(toImmutableMap(Map.Entry::getKey, entry -> {
-            ClusterMemoryPool pool = new ClusterMemoryPool(entry.getKey());
-            pool.update(memoryInfos, toIntExact(entry.getValue()));
-            return pool.getClusterInfo();
-        }));
+    private Query getLargestMemoryQuery(@Nullable Query existingLargeQuery, @NotNull Query newQuery)
+    {
+        requireNonNull(newQuery, "newQuery must not be null");
+        if (existingLargeQuery == null) {
```

Comment:
The `Nullable` and `NotNull` are very easy to mix up as they look similar from the first glance. Maybe use `Optional` to avoid annotation?

Reply:
Replaced with Optional for `existingLargestQuery`